pkg/watch/manager.go

```diff
@@ -33,330 +36,268 @@ type Manager struct {
+	u := &unstructured.Unstructured{}
+	u.SetGroupVersionKind(gvk)
+	informer, err := wm.mgr.GetCache().GetInformer(u)
+	if err != nil {
+		// This is expected to fail if a CRD is unregistered.
+		return fmt.Errorf("getting informer for kind: %+v %w", gvk, err)
 	}
+	informer.AddEventHandler(wm)
 
-	wm.watchedKinds = filteredNewWatchedKinds
-	return true, nil
+	// Mark it as watched TODO(OREN) simplify
+	wm.watchedKinds[gvk] = m[gvk]
+	return nil
 }
 
-// Start looks for changes to the watch roster every 5 seconds. This method has a
-// benefit compared to restarting the manager every time a controller changes the watch
-// of placing an upper bound on how often the manager restarts.
-func (wm *Manager) Start(done <-chan struct{}) error {
-	ticker := time.NewTicker(5 * time.Second)
-	defer ticker.Stop()
-	for {
-		select {
-		case <-done:
-			log.Info("watch manager shutting down")
-			wm.close()
-			return nil
-		case <-ticker.C:
-			if _, err := wm.updateOrPause(); err != nil {
-				log.Error(err, "error in updateManagerLoop")
-			}
-		}
-	}
+type removableCache interface {
+	Remove(obj runtime.Object) error
 }
 
-// updateOrPause() wraps the update function, allowing us to check if the manager is paused in
-// a thread-safe manner
-func (wm *Manager) updateOrPause() (bool, error) {
-	wm.startedMux.Lock()
-	defer wm.startedMux.Unlock()
-	// Report restart check after acquiring the lock so that we can detect deadlocks
-	if err := wm.metrics.reportRestartCheck(); err != nil {
-		log.Error(err, "while trying to report restart check metric")
-	}
-	if wm.paused {
-		log.Info("update manager is paused")
-		return false, nil
-	}
-	return wm.updateManager()
+func (wm *Manager) removeWatch(gvk schema.GroupVersionKind) error {
+	wm.watchedMux.Lock()
+	defer wm.watchedMux.Unlock()
+	return wm.doRemoveWatch(gvk)
 }
 
-// Pause the manager to prevent syncing while other things are happening, such as wiping
-// the data cache
-func (wm *Manager) Pause() error {
-	wm.startedMux.Lock()
-	defer wm.startedMux.Unlock()
-	if wm.stopped != nil {
-		wm.stopper()
-		select {
-		case <-wm.stopped:
-		// Choose a long enough timeout that the API server would have already timed out
-		case <-time.After(60 * time.Second):
-			return errors.New("timeout waiting for watch manager to pause")
-		}
-	}
-	wm.paused = true
-	return nil
-}
-
-// Unpause the manager and start new watches
-func (wm *Manager) Unpause() error {
-	wm.startedMux.Lock()
-	defer wm.startedMux.Unlock()
-	wm.paused = false
-	return nil
-}
+func (wm *Manager) doRemoveWatch(gvk schema.GroupVersionKind) error {
+	// lock acquired by caller
 
-// restartManager destroys the old manager and creates a new one watching the provided constraint
-// kinds
-func (wm *Manager) restartManager(kinds map[schema.GroupVersionKind]vitals) error {
-	var kindStr []string
-	for gvk := range kinds {
-		kindStr = append(kindStr, gvk.String())
+	if _, ok := wm.watchedKinds[gvk]; !ok {
```

Comment:
Do we care if we are not watching? Or should this be silent b/c mission accomplished.

Reply:
Agreed.